[lldb][lldb-dap] Added support for "WriteMemory" request.

santhoshe447 opened the pull request:
Added debug adapter support for write memory.

Commit:
[lldb][test] Add the ability to extract the variable value out of the summary.

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Added "port" property to vscode "attach" command.

Adding a "port" property to the VsCode "attach" command likely extends the functionality of the debugger configuratiuon to allow attaching to a process using PID or PORT number.
Currently, the "Attach" configuration lets the user specify a pid. We tell the user to use the attachCommands property to run "gdb-remote <port>".
Followed the below conditions for "attach" command with "port" and "pid"
We should add a "port" property. If port is specified and pid is not, use that port to attach. If both port and pid are specified, return an error saying that the user can't specify both pid and port.

Ex - launch.json
{
	"version": "0.2.0",
    "configurations": [
        {
            "name": "lldb-dap Debug",
            "type": "lldb-dap",
            "request": "attach",
            "port":1234,
            "program": "${workspaceFolder}/a.out",
            "args": [],
            "stopOnEntry": false,
            "cwd": "${workspaceFolder}",
            "env": [],

        }
    ]
}

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Added "port" property to vscode "attach" command.

Adding a "port" property to the VsCode "attach" command likely extends the functionality of the debugger configuration to allow attaching to a process using PID or PORT number.
Currently, the "Attach" configuration lets the user specify a pid. We tell the user to use the attachCommands property to run "gdb-remote ".
Followed the below conditions for "attach" command with "port" and "pid"
We should add a "port" property. If port is specified and pid is not, use that port to attach. If both port and pid are specified, return an error saying that the user can't specify both pid and port.

Ex - launch.json
{
"version": "0.2.0",
"configurations": [
{
"name": "lldb-dap Debug",
"type": "lldb-dap",
"request": "attach",
"port":1234,
"program": "${workspaceFolder}/a.out",
"args": [],
"stopOnEntry": false,
"cwd": "${workspaceFolder}",
"env": [],

    }
]
}

In this patch we have resolved code formatting issues and fixed "test_by_name" failure.

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Added "port" property to vscode "attach" command.
Resolved code format issues.

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
Resolved code format issue.

Commit:
Resolved code format errors.

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Added "port" property to vscode "attach" command - Addressed review comments
Addressed all review comments.

Commit:
[lldb-dap] Added "port" property to vscode "attach" command - fix code format
resolved code format issues.

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Added "port" property to vscode "attach" command - resolved all review comments.
All review comments have been addressed, and the attach-by-port tests have been verified on
Linux machine. Although the functionality is intended to support both Linux and macOS, we were
unable to verify it on macOS due to lack of access.

Commit:
[lldb-dap] Added "port" property to vscode "attach" command -resolved format check issues.

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Added "port" property to vscode "attach" command. #91570
Resolved few review commennts.

Commit:
[lldb-dap] Added "port" property to vscode "attach" command.
Resolved code format issues.

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Added "port" property to vscode "attach" command - addressed review comments.
Used named pipe to read debug server listening port.
Since we are only using "gdb-remote" process plugin, changed to "gdb-remote-port" and
"gdb-remote-hostname" from "port" and "hostname"

Commit:
[lldb-dap] Added "port" property to vscode "attach" command - resolved code format issues.

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Added "port" property to vscode "attach" command - fixed review comments.

Commit:
[lldb-dap] Added "port" property to vscode "attach" command - resolved review comments.

Commit:
[lldb-dap] Added "port" property to vscode "attach" command.
I have shifted "Pipe" class to common location "lldb/packages/Python/lldbsuite/test/tools/lldb-server/lldbgdbserverutils.py"

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Updated README.md for newly added attach properties.

Added "gdb-remote-port" and "gdb-remote-hostname" attach properties usage in
README.md.

This was missed in PR https://github.com/llvm/llvm-project/pull/91570

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Updated README.md for newly added attach properties. #99926
Corrected spelling mistakes.

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Bump the version to 0.2.3

Bump the lldb-dap version to 0.2.3.

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
[lldb-dap] Enabling instruction breakpoint support to lldb-dap.
Added support for "supportsInstructionBreakpoints" capability and now it this command is triggered when we set instruction breakpoint.
We need this support as part of enabling disassembly view debugging. Following features should work as part of this feature enablement:

1. Settings breakpoints in disassembly view: Unsetting the breakpoint is not happening from the disassembly view. Currently we need to unset breakpoint manually from the breakpoint List. Multiple breakpoints are getting set for the same $

2. Step over, step into, continue in the disassembly view

The format for DisassembleRequest and DisassembleResponse at https://raw.githubusercontent.com/microsoft/vscode/master/src/vs/workbench/contrib/debug/common/debugProtocol.d.ts .

santhoshe447 committed:
[lldb-dap] Enabling instruction breakpoint support to lldb-dap.
Resolved code format issues.

santhoshe447 committed:
[lldb-dap] Enabling instruction breakpoint support to lldb-dap.
Resolved code format issues.

santhoshe447 committed:
[lldb-dap] Enabling instruction breakpoint support to lldb-dap.
The following changes were committed by mistake, so they have been reverted.
"startDebugging", &g_dap.start_debugging_request_handler,
"repl-mode", &g_dap.repl_mode_request_handler,

santhoshe447 committed:
[lldb-dap] Enabling instruction breakpoint support to lldb-dap.
Few changes were missed in lldb/tools/lldb-dap/DAP.h, so added in this version.

santhoshe447 committed:
[lldb-dap] Enabling instruction breakpoint support to lldb-dap.
Resolved review comments.

santhoshe447 committed:
[lldb-dap] Enabling instruction breakpoint support to lldb-dap.
Added stop reason for instruction breakpoint.

santhoshe447 committed:
[lldb-dap] Enabling instruction breakpoint support to lldb-dap.
Added few comments

santhoshe447 committed:
Merge branch 'main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
[lldb-dap] Enabling instruction breakpoint support to lldb-dap.
Resolved follwoing review comments :
1. Removed usage of removed_ibp, removed deleted instruction breakpoints deirectly from g_dap.instruction_breakpoints.erase(prev_ibp.first)
2. Reused CreateJsonObject function in CreateInstructionBreakpoint.
3. Updated test cases with stop reason.

santhoshe447 committed:
[lldb-dap] Enabling instruction breakpoint support to lldb-dap.
Resolved code format issues.

santhoshe447 committed:
[lldb-dap] Enabling instruction breakpoint support to lldb-dap.
Resolved review comments.

santhoshe447 committed:
[lldb-dap] Enabling instruction breakpoint support to lldb-dap.
Resolved review comments.

santhoshe447 committed:
[lldb-dap] Enabling instruction breakpoint support to lldb-dap.
Resolved review comments.

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
[lldb-dap] XFailed TestDAP_instruction_breakpoint.py for windows.

TestDAP_instruction_breakpoint.py failed on windows, so these test were skipped for windows build.

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
[lldb-dap] Fix for missing 'raw_string_ostream::flush' removal in ProgressEvent.cpp; addressing #108745

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
[lldb][lldb-dap] Added WriteMemory support

Added debug adapter support for write memory.

santhoshe447 committed:
[lldb][lldb-dap] Added WriteMemory support

Added debug adapter support for write memory.

santhoshe447 committed:
Merge branch 'main' of github.com:santhoshe447/llvm-project into main

santhoshe447 committed:
[lldb][lldb-dap] Added readMemory and WriteMemory, var type correction #108036

Resolved code format issues.

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'main' of github.com:santhoshe447/llvm-project into main

santhoshe447 committed:
Merge branch 'main' of github.com:santhoshe447/llvm-project into main

JDevlieghere commented on lldb/tools/lldb-dap/CMakeLists.txt:
```diff
@@ -77,6 +77,7 @@ add_lldb_tool(lldb-dap
   Protocol/ProtocolBase.cpp
   Protocol/ProtocolTypes.cpp
   Protocol/ProtocolRequests.cpp
+  Handler/WriteMemoryRequestHandler.cpp
```
This should go between the other `Handler/` files (which are sorted alphabetically)

JDevlieghere commented on lldb/tools/lldb-dap/Handler/WriteMemoryRequestHandler.cpp:
```diff
@@ -0,0 +1,147 @@
+//===-- WriteMemoryRequestHandler.cpp
+//--------------------------------------===//
```
```suggestion
//===-- WriteMemoryRequestHandler.cpp -------------------------------------===//
```

JDevlieghere commented on lldb/tools/lldb-dap/Handler/WriteMemoryRequestHandler.cpp:
```diff
@@ -0,0 +1,147 @@
+void WriteMemoryRequestHandler::operator()(
+    const llvm::json::Object &request) const {
+  llvm::json::Object response;
+  llvm::json::Array response_writeMemory;
```
This looks unused?
```suggestion
```

JDevlieghere commented on lldb/tools/lldb-dap/Handler/WriteMemoryRequestHandler.cpp:
```diff
@@ -0,0 +1,147 @@
+    const llvm::json::Object &request) const {
+  llvm::json::Object response;
+  llvm::json::Array response_writeMemory;
+  llvm::json::Object body;
```
Move this down to where we're populating the response body (i.e. before line 141)

santhoshe447 replied:
Thanks, I have moved llvm::json::Object response;.

JDevlieghere commented on lldb/tools/lldb-dap/Handler/WriteMemoryRequestHandler.cpp:
```diff
@@ -0,0 +1,147 @@
+
+  llvm::StringRef data64 = GetString(arguments, "data").value_or("");
+
+  std::string output;
```
```suggestion
```

santhoshe447 replied:
This too

JDevlieghere commented on lldb/tools/lldb-dap/Handler/WriteMemoryRequestHandler.cpp:
```diff
@@ -0,0 +1,147 @@
+  lldb::SBError writeError;
+  int64_t countWrite = 0;
+
+  output = llvm::encodeBase64(data64);
```
```suggestion
  std::string output = llvm::encodeBase64(data64);
```

JDevlieghere commented on lldb/tools/lldb-dap/Handler/WriteMemoryRequestHandler.cpp:
```diff
@@ -0,0 +1,147 @@
+
+  std::string output;
+  lldb::SBError writeError;
+  int64_t countWrite = 0;
```
Let's go with `bytes_written` to match the name of the field in the response. Also, this should be an unsigned presumably? 
```suggestion
  uint64_t bytes_written = 0;
```

santhoshe447 replied:
I have changed this as well.

JDevlieghere commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -153,6 +153,8 @@ static void RegisterRequestCallbacks(DAP &dap) {
   // Testing requests
   dap.RegisterRequest<TestGetTargetBreakpointsRequestHandler>();
+  // WriteMemory request
+  dap.RegisterRequest<WriteMemoryRequestHandler>();
```
Similarly to the comment in the CMake file, this should get inserted into the list of regular request handlers above, which are sorted alphabetically.

santhoshe447 committed:
[lldb][lldb-dap] Added support for "WriteMemory" request. #131820

resolved review comments.

santhoshe447 committed:
[lldb][lldb-dap] Added support for "WriteMemory" request. #131820

resolved review comments.

vogelsgesang commented on lldb/tools/lldb-dap/Handler/InitializeRequestHandler.cpp:
```diff
@@ -463,6 +463,8 @@ void InitializeRequestHandler::operator()(
   body.try_emplace("supportsDataBreakpoints", true);
   // The debug adapter supports the `readMemory` request.
   body.try_emplace("supportsReadMemoryRequest", true);
+  // The debug adapter supports the 'writeMemory' request.
+  body.try_emplace("supportsWriteMemoryRequest", true);
```
with the recent refactoring from https://github.com/llvm/llvm-project/pull/131943, this capability should be declared in the `WriteMemoryRequestHandler`

santhoshe447 replied:
Hi @vogelsgesang ,
Thanks a lot for letting me know, but those latest changes are not showing to me.
I will check this, if it required, I will push another patch.

vogelsgesang commented on lldb/tools/lldb-dap/Handler/RequestHandler.h:
```diff
@@ -394,6 +394,13 @@ class TestGetTargetBreakpointsRequestHandler : public LegacyRequestHandler {
+class WriteMemoryRequestHandler : public LegacyRequestHandler {
+public:
+  using LegacyRequestHandler::LegacyRequestHandler;
+  static llvm::StringLiteral getCommand() { return "writeMemory"; }
```
recently renamed in https://github.com/llvm/llvm-project/commit/1028ea9e2641341436555347fb65d6f2eafcfbdd

```suggestion
  static llvm::StringLiteral GetCommand() { return "writeMemory"; }
```

vogelsgesang commented on lldb/tools/lldb-dap/Handler/RequestHandler.h:
```diff
@@ -394,6 +394,13 @@ class TestGetTargetBreakpointsRequestHandler : public LegacyRequestHandler {
   void operator()(const llvm::json::Object &request) const override;
 };
 
+class WriteMemoryRequestHandler : public LegacyRequestHandler {
```
I would prefer if we would directly use the non-legacy approach here

santhoshe447 commented on lldb/test/API/tools/lldb-dap/memory/TestDAP_memory.py:
```diff
@@ -112,3 +112,23 @@ def test_readMemory(self):
+        memref = ptr_deref["memoryReference"]
+
+        mem = self.dap_server.request_writeMemory(memref, 0, "0x11")["body"]
+        self.assertEqual(mem["bytesWritten"], 8)
```
Could you please elaborate this  ? Are you expecting negative test case ?

ashgti replied:
Oh, for observing the effect, we might also want to trigger a https://microsoft.github.io/debug-adapter-protocol/specification#Events_Memory event.

santhoshe447 commented on lldb/tools/lldb-dap/Handler/WriteMemoryRequestHandler.cpp:
```diff
@@ -0,0 +1,147 @@
+  output = llvm::encodeBase64(data64);
+
+  // write the memory
+  if (!output.empty()) {
```
Can you please elaborate your point?

ashgti replied:
This is checking if output is empty, which means that if I tried to set a value to `''` then it would not have an effect, as far as I can tell.

So, I'm wondering how I'm supposed to set something to `null` or empty or nil.

vogelsgesang replied:
> [...] how I'm supposed to set something to null or empty or nil.

Not sure I understand the question. But maybe this is where the confusion comes from:

This PR contains the "write memory" not the "write variable" functionality. A byte in memory / RAM always has some value (0x00, 0xff, ...), it cannot be "empty / nil". As such , the right why to set a memory location to 0 is to write a 0x00 (base64-encoded) as the target value

There are some notions of "empty" / "non-allocated" memory in address sanitizer. The Debug Adapter Protocol does not currently provide any mechanisms to interact with those sanitizer states, though

(Sorry, in case this isn't where the source of confusion here, and in case I am just stating the obvious)

santhoshe447 replied:
> > [...] how I'm supposed to set something to null or empty or nil.
> 
> Not sure I understand the question. But maybe this is where the confusion comes from:
> 
> This PR contains the "write memory" not the "write variable" functionality. A byte in memory / RAM always has some value (0x00, 0xff, ...), it cannot be "empty / nil". As such , the right why to set a memory location to 0 is to write a 0x00 (base64-encoded) as the target value
> 
> There are some notions of "empty" / "non-allocated" memory in address sanitizer. The Debug Adapter Protocol does not currently provide any mechanisms to interact with those sanitizer states, though
> 
> (Sorry, in case this isn't where the source of confusion here, and in case I am just stating the obvious)

I agree with your comment that a writeMemory request with empty data is not possible. So, I have not added this use case in the test. And the notation of Base64 string for "" is "" only.
Thanks,

santhoshe447 commented on lldb/tools/lldb-dap/Handler/WriteMemoryRequestHandler.cpp:
```diff
@@ -0,0 +1,142 @@
+  lldb::SBError writeError;
+  uint64_t bytes_written = 0;
+
+  std::string output = llvm::encodeBase64(data64);
```
oh, I will check this,

santhoshe447 committed:
Merge branch 'main' of github.com:santhoshe447/llvm-project into main

santhoshe447 committed:
Merge branch 'main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'main' of github.com:santhoshe447/llvm-project into main

santhoshe447 commented:
> There is another Pull Request for the same functionality #108036. The other PR seems to be outdated, though, and this PR seems to be further along (e.g. it also has test cases). CC @jennphilqc @walter-erquinigo @clayborg since you commented on the other PR and might also be interested in this PR

Yes, you are right. She was my colleague.
I am continuing her work and adding and adding test cases based on the new refactor changes.

santhoshe447 commented:
Hi Every,
I really appreciate the time and effort you have put into reviewing my changes. I am working on addressing all the comments accordingly.
The implementation was verified through API tests, but when I tried check it using the VSCode IDE GUI, I did not see the "writeMemory" request at all.
I have reported this to VSCode [https://github.com/microsoft/vscode/issues/244277]. It would be grateful if someone could guide me on this.

Thanks

walter-erquinigo commented:
@santhoshe447 , the VSCode folks can take up to several weeks to reply. In the meantime, you can check the vscode repo https://github.com/microsoft/vscode for usages of this request to learn more about what's going on.

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
[lldb][lldb-dap] Added support for "WriteMemory" request. #131820

Resolved all review comments.

santhoshe447 committed:
Merge branch 'main' of github.com:santhoshe447/llvm-project into main

santhoshe447 committed:
Merge branch 'main' of github.com:santhoshe447/llvm-project into main

vogelsgesang commented on lldb/tools/lldb-dap/Handler/WriteMemoryRequestHandler.cpp:
```diff
@@ -0,0 +1,175 @@
+      lldb::SBMemoryRegionInfo region_info;
+      lldb::SBError error =
+          process.GetMemoryRegionInfo(address_offset, region_info);
+      if (!error.Success() || !region_info.IsWritable()) {
+        response["success"] = false;
+        EmplaceSafeString(response, "message",
+                          "Memory 0x" + llvm::utohexstr(address_offset) +
+                              " region is not writable");
```
we are only checking if the *first byte* is writable. However, according to the DAP spec, we should be checking the complete region we are writing, i.e. if we would be able to write *all* bytes without issues. A single DAP-write might might span multiple LLDB MemoryRegionInfos.

santhoshe447 replied:
> we are only checking if the _first byte_ is writable. However, according to the DAP spec, we should be checking the complete region we are writing, i.e. if we would be able to write _all_ bytes without issues. A single DAP-write might might span multiple LLDB MemoryRegionInfos.

Just wanted to understand.
start_address = 0x100
end_address = start_address +  data.size() = > 0x104
So, do I need to verify each byte from 'start_address' to 'end_address' for writability, or should I check if the entire memory region (As given by MemoryRegionInfo -> GetRegionEnd and GetRegionBase) is writable?
From qMemoryRegionInfo, we get the start and end address of the memory region (For requested address 0x100) along with its rw permissions.
Should I rely on this to determine writability, or should I check each byte individually within the range ?

gdb-remote packet logs:

lldb             <  34> send packet: $qMemoryRegionInfo:7fffffffe528#19
lldb             <  76> read packet: $start:7ffffffde000;size:21000;permissions:rw;flags:;name:5b737461636b5d;#d9
lldb             <  22> send packet: $M7fffffffe528,1:32#4e
lldb             <   6> read packet: $OK#9a

vogelsgesang replied:
I think relying on start and end address would be fine to make sure we cover the complete address range.

However, afaik, `MemoryRegionInfo` describes the memory permissions from the point of view of the debuggee (i.e. the program being debugged). The debugger might have more permissions, i.e. maybe (?) the debugger is able to write to memory regions which for the debuggee are marked as read-only? Not sure if this case can happen, and how to properly check which regions the debugger is allowed to write to...

@labath do you know how to properly implement this check?

santhoshe447 replied:
Considering the entire region, I have modified the code. Please check and give your feedback.

```
  // Write the memory.
  if (!output.empty()) {
    lldb::SBProcess process = dap.target.GetProcess();
    // If 'allowPartial' is false or missing, a debug adapter should attempt to
    // verify the region is writable before writing, and fail the response if it
    // is not.
    if (!allowPartial) {
      // Compute the end of the write range.
      lldb::addr_t end_address = address + output.size();
      // Get memory region info for the given address.
      // This provides the region's base, end, and permissions
      // (read/write/executable).
      lldb::SBMemoryRegionInfo region_info;
      lldb::SBError error = process.GetMemoryRegionInfo(address, region_info);
      // Fail if the region info retrieval fails, is not writable, or the range
      // exceeds the region.
      if (!error.Success() || !region_info.IsWritable() ||
          end_address > region_info.GetRegionEnd()) {
        dap.SendErrorResponse(response, "Memory 0x" + llvm::utohexstr(address) +
                                            " region is not writable");
        return;
      }
    }

    bytes_written =
        process.WriteMemory(address, static_cast<void *>(output.data()),
                            output.size(), write_error);
```

vogelsgesang commented on lldb/test/API/tools/lldb-dap/memory/TestDAP_memory.py:
```diff
@@ -112,3 +112,48 @@ def test_readMemory(self):
+        self.assertEqual(mem_response["body"]["bytesWritten"], 1)
+
+        # Read back the modified memory and verify that the written data matches
+        # the expecte result.
```
```suggestion
        # the expected result.
```

vogelsgesang commented on lldb/test/API/tools/lldb-dap/memory/TestDAP_memory.py:
```diff
@@ -112,3 +112,48 @@ def test_readMemory(self):
+        )
+        self.continue_to_next_stop()
+
+        # Get the 'not_a_ptr' writable variablle reference address.
```
```suggestion
        # Get the 'not_a_ptr' writable variable reference address.
```

vogelsgesang commented on lldb/test/API/tools/lldb-dap/memory/TestDAP_memory.py:
```diff
@@ -112,3 +112,48 @@ def test_readMemory(self):
+        memref = ptr_deref["memoryReference"]
+
+        # Write the Base64-encoded string "Mg==", which decodes to binary 0x32
+        # which is decimal 50 and corresponds to the SCII character '2'.
```
```suggestion
        # which is decimal 50 and corresponds to the ASCII character '2'.
```

vogelsgesang commented on lldb/test/API/tools/lldb-dap/memory/TestDAP_memory.py:
```diff
@@ -112,3 +112,48 @@ def test_readMemory(self):
+        ptr_deref = self.dap_server.request_evaluate("nonWritable")["body"]
+        memref = ptr_deref["memoryReference"]
+
+        # Writing to non-writable region should return an appropriate error.
+        mem_response = self.dap_server.request_writeMemory(memref, 0, "Mg==", False)
+        self.assertEqual(mem_response["success"], False)
```
actually verify the error message

ashgti commented on lldb/test/API/tools/lldb-dap/memory/TestDAP_memory.py:
```diff
@@ -112,3 +112,48 @@ def test_readMemory(self):
+        ptr_deref = self.dap_server.request_evaluate("not_a_ptr")["body"]
+        memref = ptr_deref["memoryReference"]
+
+        # Write the Base64-encoded string "Mg==", which decodes to binary 0x32
```
Instead of a magic variable, you could add a helper to `lldbdap_testcase.py` that uses the [base64](https://docs.python.org/3/library/base64.html) python module to encode byte strings.
E.g. your example is:
```
$ python3
>>> import base64
>>> base64.b64encode(b'2')
b'Mg=='
```

And it might be a bit more clear to the reader whats happening.

Something along the lines of:
```
def writeMemory(self, memory_reference, data, offset=None, allow_partial=None):
   encodedData = base64.b64encode
   resp = self.dap_server.request_writeMemory(memory_reference, offset=offset, allow_partial=allow_partial, data=encodedData)
   self.assertTrue(resp['success'])
   self.assertEqual(resp['body']['bytesWritten'], len(data))
```

santhoshe447 replied:
> Instead of a magic variable, you could add a helper to `lldbdap_testcase.py` that uses the [base64](https://docs.python.org/3/library/base64.html) python module to encode byte strings. E.g. your example is:
> 
> ```
> $ python3
> >>> import base64
> >>> base64.b64encode(b'2')
> b'Mg=='
> ```
> 
> And it might be a bit more clear to the reader whats happening.
> 
> Something along the lines of:
> 
> ```
> def writeMemory(self, memory_reference, data, offset=None, allow_partial=None):
>    encodedData = base64.b64encode
>    resp = self.dap_server.request_writeMemory(memory_reference, offset=offset, allow_partial=allow_partial, data=encodedData)
>    self.assertTrue(resp['success'])
>    self.assertEqual(resp['body']['bytesWritten'], len(data))
> ```

Thanks for your suggestion.
I will update accordingly.

ashgti replied:
Or you can use `base64.b64encode((50).to_bytes())` to get the `Mg==` value

santhoshe447 committed:
Merge branch 'main' of github.com:santhoshe447/llvm-project into main

santhoshe447 committed:
[lldb][lldb-dap] Added support for "WriteMemory" request. #131820
Fixed review comments

clayborg commented on lldb/tools/lldb-dap/Handler/WriteMemoryRequestHandler.cpp:
```diff
@@ -0,0 +1,175 @@
+  lldb::addr_t address_offset =
+      address + GetInteger<uint64_t>(arguments, "offset").value_or(0);
+
+  llvm::StringRef data64 = GetString(arguments, "data").value_or("");
```
Do we want to return an error if data64 is empty here? Would be good to test no data or an empty data in the tests. Or will the `decodeError` error below catch this? If the "data" argument is not there or empty, we should return an appropriate error, otherwise we will fall through to below and we will return `writeError.GetCString()` as the error, but writing zero bytes might not return an error

santhoshe447 replied:
I have added additional checks to validate empty data and added this case in test.

clayborg commented on lldb/tools/lldb-dap/Handler/WriteMemoryRequestHandler.cpp:
```diff
@@ -0,0 +1,175 @@
+  // This function decodes it into raw binary before writing it to the target
+  // process memory.
+  std::vector<char> output;
+  auto decodeError = llvm::decodeBase64(data64, output);
```
s/decodeError/decode_error/

clayborg commented on lldb/tools/lldb-dap/Handler/WriteMemoryRequestHandler.cpp:
```diff
@@ -0,0 +1,175 @@
+  auto decodeError = llvm::decodeBase64(data64, output);
+
+  if (decodeError) {
+    response["success"] = false;
+    EmplaceSafeString(response, "message",
+                      llvm::toString(std::move(decodeError)).c_str());
+    dap.SendJSON(llvm::json::Value(std::move(response)));
```
It would be nice to make a utility function to return an error in the response. We see this kind of "set success to false, make a string, send the packet. 
```
if (decode_error)
  return dap.SendError(response, llvm::toString(std::move(decodeError)).c_str())
```
It would clean up a lot of code. We need to provide the response that has the packet ID and other stuff already encoded.

vogelsgesang replied:
we already have utility classes for this. Hence also my request to use the "new" instead of the "legacy" handler infrastructure. See https://github.com/llvm/llvm-project/pull/131820#discussion_r2003256423

santhoshe447 replied:
As of now I have overloaded EmplaceSafeString to accept th esuccess status and dap instance.
Would it be fine to overload "EmplaceSafeString" api ?

santhoshe447 replied:
Added new utility function in DAP.h
void DAP::SendErrorResponse(llvm::json::Object &response,
                            llvm::StringRef message)

clayborg commented on lldb/tools/lldb-dap/Handler/WriteMemoryRequestHandler.cpp:
```diff
@@ -0,0 +1,175 @@
+  }
+
+  lldb::addr_t address = *addr_opt;
+  lldb::addr_t address_offset =
```
`address_offset` seems to indicate this is an offset. Maybe use `address_plus_offset`, or just use `address`:
```
lldb::addr_t address = *addr_opt + GetInteger<uint64_t>(arguments, "offset").value_or(0);
```

santhoshe447 replied:
> `address_offset` seems to indicate this is an offset. Maybe use `address_plus_offset`, or just use `address`:
> 
> ```
> lldb::addr_t address = *addr_opt + GetInteger<uint64_t>(arguments, "offset").value_or(0);
> ```

Thanks, I have simplified this.

clayborg commented on lldb/tools/lldb-dap/Handler/WriteMemoryRequestHandler.cpp:
```diff
@@ -0,0 +1,175 @@
+    return;
+  }
+
+  bool allowPartial = GetBoolean(arguments, "allowPartial").value_or(true);
```
`allow_partial`

santhoshe447 replied:
> `allow_partial`

I have adopted your suggested naming convention for some variables. However, I kept the argument names the same as received from DAP, as suggested in the review comments. I believe this helps maintain better readability.

clayborg commented on lldb/tools/lldb-dap/Handler/WriteMemoryRequestHandler.cpp:
```diff
@@ -0,0 +1,175 @@
+  }
+
+  bool allowPartial = GetBoolean(arguments, "allowPartial").value_or(true);
+  lldb::SBError writeError;
```
`write_error`

santhoshe447 replied:
> `write_error`

changed to write_error

clayborg commented on lldb/tools/lldb-dap/Handler/WriteMemoryRequestHandler.cpp:
```diff
@@ -0,0 +1,175 @@
+  FillResponse(request, response);
+
+  auto arguments = request.getObject("arguments");
+  llvm::StringRef memoryReference =
+      GetString(arguments, "memoryReference").value_or("");
+
+  auto addr_opt = DecodeMemoryReference(memoryReference);
+  if (!addr_opt.has_value()) {
+    response["success"] = false;
+    response["message"] =
+        "Malformed memory reference: " + memoryReference.str();
+    dap.SendJSON(llvm::json::Value(std::move(response)));
+    return;
+  }
```
Can we add a utility function for decoding memory references? This code is going to be duplicated everywhere that uses memory references. Maybe something like:
```
std::optional<lldb::addr_t> memory_reference = GetMemoryReference(arguments, "memoryReference");
```

santhoshe447 replied:
> Can we add a utility function for decoding memory references? This code is going to be duplicated everywhere that uses memory references. Maybe something like:
> 
> ```
> std::optional<lldb::addr_t> memory_reference = GetMemoryReference(arguments, "memoryReference");
> ```

We can add a utility function, but it may not be very useful in all cases.
Sometimes, we need memoryReference in string format, while other times, we need it as addr_t.
Given these variations, I don't think having a separate utility function would be beneficial, and we won't reduce significant code duplicate.
Still If you recommend, I will push it in different pull request.

clayborg commented on lldb/tools/lldb-dap/Handler/WriteMemoryRequestHandler.cpp:
```diff
@@ -0,0 +1,175 @@
+  FillResponse(request, response);
+
+  auto arguments = request.getObject("arguments");
+  llvm::StringRef memoryReference =
```
`memory_reference`

santhoshe447 replied:
> `memory_reference`

I have adopted your suggested naming convention for some variables. However, I kept the argument names the same as received from DAP, as suggested in the review comments. I believe this helps maintain better readability.

santhoshe447 replied:
I have updated this to memory_reference from memoryReference.

santhoshe447 committed:
Merge branch 'main' of github.com:santhoshe447/llvm-project into main

santhoshe447 committed:
[lldb][lldb-dap] Added support for "WriteMemory" request. #131820
Added separate utility function to send error response

santhoshe447 committed:
Merge branch 'main' of github.com:santhoshe447/llvm-project into main

santhoshe447 committed:
Merge branch 'main' of github.com:santhoshe447/llvm-project into main

santhoshe447 committed:
Merge branch 'main' of github.com:santhoshe447/llvm-project into main

santhoshe447 committed:
[lldb][lldb-dap] Added support for "WriteMemory" request. #131820

Resolved review comments and added additional case in writeMemory test.

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
[lldb][lldb-dap] Added support for "WriteMemory" request. #131820

Adopted latest RequestHandler changes for WriteMemory.

santhoshe447 committed:
Merge branch 'main' into main

santhoshe447 committed:
[lldb][lldb-dap] Added support for "WriteMemory" request. #131820

Resolved code format issues.

santhoshe447 committed:
Merge branch 'llvm:main' into main

tedwoodward commented on lldb/tools/lldb-dap/Handler/WriteMemoryRequestHandler.cpp:
```diff
@@ -0,0 +1,175 @@
+    // is not.
+    if (!allowPartial) {
+      // Compute the end of the write range.
+      lldb::addr_t end_address = address + output.size();
```
should be address + output.size() -1, since size 4 starting at 1000 goes from 1000 - 1003.

santhoshe447 replied:
Thanks, changed this.

tedwoodward commented on lldb/tools/lldb-dap/Handler/WriteMemoryRequestHandler.cpp:
```diff
@@ -0,0 +1,175 @@
+      lldb::SBError error = process.GetMemoryRegionInfo(address, region_info);
+      // Fail if the region info retrieval fails, is not writable, or the range
+      // exceeds the region.
+      if (!error.Success() || !region_info.IsWritable() ||
```
The write might spill into the next region (or more). Say we write 0x40 bytes starting at 0xff0, and we have regions from 0x0-0xfff and 0x1000-0x1fff. If both are writeable, we want to be able to write. So if the end address is > than the region end address, get the next region (region end address + 1), and test for writeable, and the end address against the 2nd region's end address. repeat until success (writeable and end < region end), or failure (not writeable).
What to do if there is no region is debatable.

santhoshe447 replied:
Thanks @tedwoodward,
Yeah, I understood your point. I have pushed the updated code to achieve the expectations.

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
[lldb][lldb-dap] Added support for "WriteMemory" request. #131820

Addressed review comments related "alloPartial = false".

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
[lldb][lldb-dap] Added support for "WriteMemory" request. #156761
resolved format error issues.

santhoshe447 committed:
[lldb][lldb-dap] Added support for "WriteMemory" request. #157136
Resolved code format issues.

santhoshe447 committed:
[lldb][lldb-dap] Added support for "WriteMemory" request. #131820

Move to the start of the next memory region + 1.

santhoshe447 commented:
Thank you so much for your valuable feedback and suggestions. I have incorporated most of the requested changes, with just two remaining suggestions that I will address in a follow-up PR.
1. Moving to non-legacy approach to handle memoryWrite/memoryWrite.
2. Adding a utility function for memoryReference argument.

Would you be kind enough to review the current changes at your convenience time? Please suggest if I am missing anything to mention here

Thanks & Best regards,

santhoshe447 committed:
Merge branch 'main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 commented:
Hi, any further comments or suggestion on the changes.

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

ashgti commented on lldb/tools/lldb-dap/DAP.h:
```diff
@@ -416,6 +416,14 @@ struct DAP {
 
   lldb::SBMutex GetAPIMutex() const { return target.GetAPIMutex(); }
 
+  /// Sends an error response in DAP with success=false and an error message.
+  ///
+  /// \param[out] response
+  ///   The response object to update.
+  /// \param[in] message
+  ///   The error message.
+  void SendErrorResponse(llvm::json::Object &response, llvm::StringRef message);
+
```
Could we move this helper to the LegacyRequestHandler? I think with the new `RequestHandler<>` type we shouldn't need this and having it on the DAP object may add new usages of this before we can finish migrating the request handlers to the new `RequestHandler<>` type.

santhoshe447 committed:
Resolved merge conflicts.
Merge remote-tracking branch 'upstream/main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
[lldb][lldb-dap] Added support for "WriteMemory" request.
Followed non-legacy approach to implement "writeMemory" request.

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'main' into main

ashgti reviewed (approved):
Looks good to me, but please fix the CI checks / formatting

santhoshe447 committed:
Update ProtocolRequests.cpp

Resolved conflicts and format issues
